daemon/containerd/image_list.go

```diff
@@ -282,7 +273,7 @@ func (i *ImageService) singlePlatformImage(ctx context.Context, contentStore con
 		Created:     rawImg.CreatedAt.Unix(),
 		RepoDigests: repoDigests,
 		RepoTags:    repoTags,
-		Size:        totalSize,
+		Size:        unpackedUsage.Size,
```

Comment:
IMO, behavior before this PR was correct, as we now store more things (packed and unpacked content). It also matched the godoc:

https://github.com/moby/moby/blob/3a520c1c0f0837e652347bca5d4f27ee0fd99a2f/api/types/image_summary.go#L80-L83

Reply:
`TestDiskUsage/after_LoadBusybox` says otherwise.

> It also matched the godoc:

I don't understand what you mean, the go doc only talks about layers, so the code only now matches the go doc

Reply:
> I don't understand what you mean, the go doc only talks about layers, so the code only now matches the go doc

**Total** size of the image **including** all layers it is composed of.

To me it sounds explicitly that this is the **total** size, so all the things associated with the image, **including** all layers, so the size ALSO includes layer size.

> `TestDiskUsage/after_LoadBusybox` says otherwise.

Isn't that because graphdrivers only stored the unpacked content, so at the time it was written it couldn't really account for something that it didn't store?
Also, packed content is also stored on disk. Why `DiskUsage` API call should only count subset of the space used by the image?

Reply:
(I should move this comment to a ticket as well) Yes, this was part of the discussions we had, which boiled down to a couple of things;

- Currently, the size we show is (roughly) "amount of space taken by this image". We may be missing some information (I guess we never took the manifest size into account as that used to be "minimal"), but generally, it's the size taken by the image.
- With graphdrivers, that is the size of the _extracted_ layers, because in some cases that's the only thing that exists (image created locally), and for cases where we pulled (or imported) the image, the compressed artefacts are discarded (as they're not needed locally)
- (small side-step: Graphdrivers considered this to be the image, because it's the only reproducible representation of images, separate from any compression to distribute the image (through registries etc).
-  This information is useful to learn "how much space is taken by (an) image(s)"
- Not so useful to learn "how large is the image if I would push (or pull) the image", which has led to confusion in the past; https://github.com/docker/hub-feedback/issues/331

With containerd, there's more than 1 representation of an image; and sometimes _multiple_ for the same image _on the same machine_;


- compressed layers and manifest(s) as distributed (pushed/pulled)
- extracted layers (snapshot)
- ☝️ which can _technically_ be multiple (multiple containers using different snapshotters, but the same image)
- 1 or more architectures (graph drives would only keep 1, but containerd can keep multiple for the same image)

As is clear, the possible list of "sizes" is diverse, and _depending on context_ some may be relevant in some contexts, and some in other, such as;

- "how much data did I pull?" or "will I push"?
- "how much disk space can I clean by deleting this image?"  (as a whole)
- "how much disk space can I prune by deleting unused platforms"?

We don't have that level of granularity (yet) in the API and other places, so those all need to be designed, and implemented.

There's also one part that is somewhat tricky: should snapshots be accounted to the _image_ or _container_? If there's no container running an image, we wouldn't need a snapshot (?). But then again, that could be extended to "we also wouldn't need the content (compressed layers)", so I guess it's somewhere in between.

However, pending any changes, we need to work with "what we have", and we should probably keep the status quo: "Size" is currently interpreted as the total size that can be accounted to the image; while we didn't include the manifest in the graph driver situation, I would consider that to be an "omission / oversight". And given that the size reported in the snapshotter situation _already_ will differ (I guess?) we could include it. So to show "all space taken up", I guess this would be the sum of;

- manifest(s)
- layers (compressed)
- snapshot(s)
- more(?)

Reply:
> Why `DiskUsage` API call should only count subset of the space used by the image?

I guess so that we are compatible as much as we can with what's existing, then later we can think about what's an image size and what kind of information do we want to give to the user

Reply:
> Isn't that because graphdrivers only stored the unpacked content, so at the time it was written it couldn't really account for something that it didn't store?

Whatever the reason for the current implementation had to show this piece of information, IMO we should make it so that the switch to containerd is the least disruptive.